### PR TITLE
Write equations directly to file

### PR DESCRIPTION
**Pull request by sjoelund**
Some refactoring to make Susan able to write some code directly to
file instead of keeping temporary strings in memory.

### CODE REVIEW

**Comment by OpenModelica-Hudson** [comment]
The test suite is unstable according to [OpenModelica_TEST_PULL_REQUEST 2016-05-27_12-57-11](https://test.openmodelica.org/hudson//job/OpenModelica_TEST_PULL_REQUEST/2836/).

**Comment by OpenModelica-Hudson** [comment]
The test suite is unstable according to [OpenModelica_TEST_PULL_REQUEST 2016-05-27_13-11-59](https://test.openmodelica.org/hudson//job/OpenModelica_TEST_PULL_REQUEST/2837/).

File: Compiler/Template/CodegenC.tpl [hunk 1]
```diff
@@ -3827,7 +3809,9 @@ template evaluateDAEResiduals(list<list<SimEqSystem>> resEquations, String model
     data->simulationInfo->callStatistics.functionEvalDAE++;
 
     <%if Flags.isSet(Flags.PARMODAUTO) then 'PM_functionDAERes(<%nrfuncs%>, data, threadData, functionDAERes_systems);'
-    else '<%fncalls%>' %>
+    else (resEquations |> eqLst => (eqLst |> eq hasindex i0 =>
+                    equation_call(eq, modelNamePrefix)
+                    ;separator="\n");separator="\n") %>
```
Comment (hkiel) [inline Compiler/Template/CodegenC.tpl]: @sjoelund I guess this part of the patch is not correct.
with old code buffer eqns and eqFuncs were filled. In line 3820 you do not remove printing of eqFunc ( which is empty at that point, so no problem, but superfluous) Here (3830), buffer fncalls was printed - but you put content of old buffer eqns...

File: Compiler/Template/CodegenC.tpl [hunk 2]
```diff
@@ -5304,32 +5289,21 @@ case SES_IFEQUATION(ifbranches=ifbranches, elsebranch=elsebranch) then
...
     }
     >>
     ;separator="\n")
-  let elseequations = ( elsebranch |> eqn =>
-     let eqnStr = equation_(-1, eqn, context, &eqnsDecls /*EQNBUF*/, modelNamePrefixStr)
-       <<
-       <%eqnStr%>
-       >>
-    ;separator="\n")
+  let &eqnsDecls += ( elsebranch |> eqn => equation_impl(-1, eqn, context, modelNamePrefixStr) ; separator="\n" )
```
Comment (hkiel) [inline Compiler/Template/CodegenC.tpl]: the returned buffer is never used

**Comment by hkiel** [comment]
Maybe numbering is off. In the following diff of the generated files you see a constant offset of 5 for both called functions:
```
diff -BurN old/dynamicTearing1.c new/dynamicTearing1.c
--- old/dynamicTearing1.c       2016-06-09 15:17:13.000000000 +0200
+++ new/dynamicTearing1.c       2016-06-09 14:33:18.000000000 +0200
@@ -277,9 +277,9 @@
   data->simulationInfo->needToIterate = 0;
   data->simulationInfo->discreteCall = 1;
   dynamicTearing1_functionLocalKnownVars(data, threadData);
-  dynamicTearing1_eqFunction_23(data, threadData);
+  dynamicTearing1_eqFunction_18(data, threadData);
 
-  dynamicTearing1_eqFunction_33(data, threadData);
+  dynamicTearing1_eqFunction_28(data, threadData);
 
   dynamicTearing1_eqFunction_34(data, threadData);

```

File: Compiler/Template/CodegenC.tpl [hunk 3]
```diff
@@ -4842,14 +4814,27 @@ template equation_(Integer clockIndex, SimEqSystem eq, Context context, Text &eq
...
+ "Generates an equation.
+  This template should not be used for a SES_RESIDUAL.
+  Residual equations are handled differently."
+::=
+  match eq
+  case e as SES_ALGORITHM(statements={})
+  then ""
+  else
+  (
+  let ix = equationIndex(eq)
```
Comment (hkiel) [inline Compiler/Template/CodegenC.tpl]: @sjoelund This line must be replaced to handle dynamic tearing correctly:
```
  let ix = match eq
  case e as SES_LINEAR(alternativeTearing = SOME(LINEARSYSTEM))
  case e as SES_NONLINEAR(alternativeTearing = SOME(NONLINEARSYSTEM)) then
    equationIndexAlternativeTearing(eq)
  else
    equationIndex(eq)
```

together with proposed change in line 3830 the dynamicTearing1 test runs fine again

Here, dynamicTearing2 still fails, but that one also fails for a pristine check out (stupid OS X?) :-(

**Comment by hkiel** [comment]
@sjoelund have you seen the pull request on your fork? https://github.com/sjoelund/OMCompiler/pull/1
That should fix this PR.

**Comment by OpenModelica-Hudson** [comment]
The tests run correctly according to [OpenModelica_TEST_PULL_REQUEST 2016-06-15_10-45-16](https://test.openmodelica.org/hudson//job/OpenModelica_TEST_PULL_REQUEST/2921/console).

**Pull request merged**